### PR TITLE
Supply denominators

### LINKED ISSUE #109846: Denominator not set for apiserver_flowcontrol_read_vs_write_request_count_samples
### What happened?

While working on the code, I noticed a bug: apiserver_flowcontrol_read_vs_write_request_count_samples is a histogram of _ratios_, but when API Priority and Fairness is used the denominator is left defaulted to 1.0 rather than set to something meaning like what is done when the max-in-flight filters are used.

Same for apiserver_flowcontrol_read_vs_write_request_count_watermarks.

### What did you expect to happen?

I expected the values observed to be in appropriate correspondence with the buckets, which are designed for observations of utilization rather than absolute numbers.

### How can we reproduce it (as minimally and precisely as possible)?

In the code, see how the call to set the denominator is only done when the max-in-flight filter is used.

For an example of bad data, look at the prometheus snapshot in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/109729/pull-kubernetes-e2e-gce-100-performance/1522280457981399040, for example at 2022-05-05 19:16:33.  Following is what I see (from a Prometheus server's web UI).

apiserver_flowcontrol_read_vs_write_request_count_samples_bucket{endpoint="apiserver", instance="10.40.0.3:443", job="master", le="+Inf", namespace="monitoring", phase="waiting", request_kind="mutating", service="master"} | 1499853
-- | --
apiserver_flowcontrol_read_vs_write_request_count_samples_bucket{endpoint="apiserver", instance="10.40.0.3:443", job="master", le="0", namespace="monitoring", phase="waiting", request_kind="mutating", service="master"} | 1494565
apiserver_flowcontrol_read_vs_write_request_count_samples_bucket{endpoint="apiserver", instance="10.40.0.3:443", job="master", le="0.1", namespace="monitoring", phase="waiting", request_kind="mutating", service="master"} | 1494565
apiserver_flowcontrol_read_vs_write_request_count_samples_bucket{endpoint="apiserver", instance="10.40.0.3:443", job="master", le="0.2", namespace="monitoring", phase="waiting", request_kind="mutating", service="master"} | 1494565
apiserver_flowcontrol_read_vs_write_request_count_samples_bucket{endpoint="apiserver", instance="10.40.0.3:443", job="master", le="0.3", namespace="monitoring", phase="waiting", request_kind="mutating", service="master"} | 1494565
apiserver_flowcontrol_read_vs_write_request_count_samples_bucket{endpoint="apiserver", instance="10.40.0.3:443", job="master", le="0.4", namespace="monitoring", phase="waiting", request_kind="mutating", service="master"} | 1494565
apiserver_flowcontrol_read_vs_write_request_count_samples_bucket{endpoint="apiserver", instance="10.40.0.3:443", job="master", le="0.5", namespace="monitoring", phase="waiting", request_kind="mutating", service="master"} | 1494565
apiserver_flowcontrol_read_vs_write_request_count_samples_bucket{endpoint="apiserver", instance="10.40.0.3:443", job="master", le="0.6", namespace="monitoring", phase="waiting", request_kind="mutating", service="master"} | 1494565
apiserver_flowcontrol_read_vs_write_request_count_samples_bucket{endpoint="apiserver", instance="10.40.0.3:443", job="master", le="0.7", namespace="monitoring", phase="waiting", request_kind="mutating", service="master"} | 1494565
apiserver_flowcontrol_read_vs_write_request_count_samples_bucket{endpoint="apiserver", instance="10.40.0.3:443", job="master", le="0.8", namespace="monitoring", phase="waiting", request_kind="mutating", service="master"} | 1494565
apiserver_flowcontrol_read_vs_write_request_count_samples_bucket{endpoint="apiserver", instance="10.40.0.3:443", job="master", le="0.9", namespace="monitoring", phase="waiting", request_kind="mutating", service="master"} | 1494565
`apiserver_flowcontrol_read_vs_write_request_count_samples_bucket{endpoint="apiserver", instance="10.40.0.3:443", job="master", le="1", namespace="monitoring", phase="waiting", request_kind="mutating", service="master"} | 1499547

### Anything else we need to know?

_No response_

### Kubernetes version

<details>
This is a long-standing bug.
</details>

### Cloud provider

<details>
N/A
</details>

### OS version

<details>
N/A
</details>

### Install tools

<details>
N/A
</details>

### Container runtime (CRI) and version (if applicable)

<details>
N/A
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
N/A
</details>

### LINKED ISSUE #110160: Wrong Denominator for priority_level_request_count_samples waiting requests
### What happened?

While working on #110101 and related PRs, I noticed that the denominator that is set for the number of waiting requests of a priority level is wrong: it is the length limit for a single queue rather than the collective limit for all the priority level's queues.

https://github.com/kubernetes/apiserver/blob/v0.24.0/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go#L242-L246

The corresponding histogram buckets assume the values are in the range [0, 1].

https://github.com/kubernetes/apiserver/blob/v0.24.0/pkg/util/flowcontrol/metrics/metrics.go#L131-L149

### What did you expect to happen?

I expected the histogram buckets to be consistent with the values observed.

### How can we reproduce it (as minimally and precisely as possible)?

See the cited code.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

Latest sources, and it has been this way for a long time.

</details>

### Cloud provider

<details>
N/A
</details>

### OS version

<details>

N/A

</details>

### Install tools

<details>
N/A
</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
This PR fixes the metric denominator problems noted in issues #109846 and #110160 .
There is more APF metrics cleanup to follow in later PRs.

#### Which issue(s) this PR fixes:

Fixes #109846
Fixes #110160

#### Does this PR introduce a user-facing change?

```release-note
The `priority_level_request_utilization` metric histogram is adjusted so that for the cases where `phase=waiting` the denominator is the cumulative capacity of all of the priority level's queues.
The `read_vs_write_current_requests` metric histogram is adjusted, in the case of using API Priority and Fairness instead of max-in-flight, to divide by the relevant limit: sum of queue capacities for waiting requests, sum of seat limits for executing requests.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/sig api-machinery
/sig instrumentation
/cc @dgrisonnet 
/cc @wojtek-t 
@tkashem 
@lavalamp

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
@cyang49

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
[the remark this answers is not in the document]
Comment (yangjunmyfm192085) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:2]: The orignal file  has changed. Does this need to be changed to 2022?

**Comment by wojtek-t** [comment]
@MikeSpreitzer - can you please rebase (based on the already merged PR)?

**Comment by MikeSpreitzer** [comment]
The latest force-push is a rebase onto `master`, now that that has all the pre-reqs.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go [hunk 2]
```diff
@@ -162,6 +164,9 @@ type configController struct {
 	// name to the state for that level.  Every name referenced from a
 	// member of `flowSchemas` has an entry here.
 	priorityLevelStates map[string]*priorityLevelState
+
+	// totalLimitConsumers are to be kept appraised of the overall limits
+	totalLimitConsumers []func(maxWaitingRequests, maxExecutingRequests int)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:169]: I think I'm still struggling with understand why we need to complicate the code so much for metrics.

What I would do is simply in lines 546-548 hardcode the metrics (which are P&F metrics anyway) and do something like:
```
fcmetrics.ReadWriteConcurrencyGaugeVec.NewForLabelValuesSafe(0, 1, []string{fcmetrics.LabelValueWaiting, epmetrics.ReadOnlyKind}),SetDenominator(meal.MaxWaiting)
fcmetrics.ReadWriteConcurrencyGaugeVec.NewForLabelValuesSafe(0, 1, []string{fcmetrics.LabelValueWaiting, epmetrics.MutatingKind}).SetDenominator(meal.MaxWaiting)
(and same for mutating)
```

I think that passing thing around, all those fields etc. is making this code very hard to follow...

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:169]: The problem is that lines 546--548 in apf_controller.go need to do what is primarily an up-call, to the higher layer in priority-and-fairness.go that updates the watermark data structures there and in maxinflight.go.  You have proposed a solution that folds in the knowledge of how those watermark data structures are composed, lessening the modularity.  That is not a deal-breaker, but something to consider.

I am struggling to remember why apf_filter.go:Interface allows registration of a multiplicity of request limit consumers; I think only one is needed.

Note that every call to NewForLabelValuesSafe returns a distinct object --- with its own denominator variable.

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:169]: The vector structure there is not really needed.   We statically know that it has exactly four members.  We can replace the vector with four individuals; that will eliminate the calls to NewForLabelValuesSafe with their attendant new object every time.

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:169]: No, wait, we need to use a vector so that all four time series show up as one metric with different label sets.

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go]: I revised, using a combination of the not-incorrect ideas above.

File: staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go [hunk 3]
```diff
@@ -78,6 +76,9 @@ func WithPriorityAndFairness(
 		klog.Warningf("priority and fairness support not found, skipping")
 		return handler
 	}
+	waitingMark.readOnlyObserver = fcmetrics.GetReadOnlyWaitingConcurrency()
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:79]: This seems to be causing a race in our tests:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/110164/pull-kubernetes-unit/1550364149416988672

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:79]: Oh, I did not think that writes of the same value would be a problem.  Easily fixed...

File: staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go [hunk 4]
```diff
@@ -108,14 +102,29 @@ func startWatermarkMaintenance(watermark *requestWatermark, stopCh <-chan struct
...
-	// Periodically observe the watermarks. This is done to ensure that they do not fall too far behind. When they do
-	// fall too far behind, then there is a long delay in responding to the next request received while the observer
-	// catches back up.
-	go wait.Until(func() {
-		watermark.readOnlyObserver.Add(0)
-		watermark.mutatingObserver.Add(0)
-	}, observationMaintenancePeriod, stopCh)
+var initMaxinflightOnce sync.Once
+
+func initMaxinflight(nonMutatingLimit, mutatingLimit int) {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go:109]: nit: for consistency with other places in this file: `initMaxInFlight` ?

File: staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go [hunk 5]
```diff
@@ -108,14 +102,29 @@ func startWatermarkMaintenance(watermark *requestWatermark, stopCh <-chan struct
...
+	initMaxinflightOnce.Do(func() {
+		// Fetching these gauges is delayed until after their underlying metric has been registered
+		// so that this latches onto the efficient implementation.
+		watermark.readOnlyObserver = fcmetrics.GetExecutingReadonlyConcurrency()
+		watermark.mutatingObserver = fcmetrics.GetExecutingMutatingConcurrency()
+		if nonMutatingLimit != 0 {
+			watermark.readOnlyObserver.SetDenominator(float64(nonMutatingLimit))
+			klog.V(2).InfoS("Set denominator for readonly requests", "limit", nonMutatingLimit)
+		} else {
+			klog.V(2).InfoS("Did not set denominator for readonly requests")
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go:119]: I suggest removing this - the lack of log is also an information. (Same below).

File: staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go [hunk 6]
```diff
@@ -132,14 +141,19 @@ func WithMaxInFlightLimit(
...
 	if mutatingLimit != 0 {
 		mutatingChan = make(chan bool, mutatingLimit)
-		watermark.mutatingObserver.SetDenominator(float64(mutatingLimit))
+		klog.V(2).InfoS("Initialized mutatingChan", "len", mutatingLimit)
+	} else {
+		klog.V(2).InfoS("Running with nil mutatingChan")
 	}
 
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		initMaxinflight(nonMutatingLimit, mutatingLimit)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go:156]: I know that synchronization on `once` is relatively cheap, but still I think we should avoid that in the handler function.
Why not calling it directly from WithMaxInFlightLimit - all metrics should be initialized at this point anyway.

File: staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go [hunk 7]
```diff
@@ -79,6 +80,13 @@ func WithPriorityAndFairness(
 		return handler
 	}
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		initAPFOnce.Do(func() {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:83]: Same comment as for max-in-flight.

File: staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go (hunk 6 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go:156]: I was unsure of that.
I was not confident that this piece of code should rely on that knowledge.

**Comment by MikeSpreitzer** [comment]
The force-pushes to 631c6950eae and fdd921cad0c make the changes suggested above and add logging of whether the efficient or inefficient case happened.

File: staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go (hunk 4 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go]: done

File: staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go (hunk 5 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go]: done

File: staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go (hunk 6 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go]: done

File: staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go (hunk 7 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go]: done

**Comment by wojtek-t** [comment]
/lgtm
/approve

Thanks!
